crates/uv-distribution/src/metadata/dependency_groups.rs

```diff
@@ -0,0 +1,193 @@
+use std::collections::BTreeMap;
+use std::path::Path;
+
+use uv_configuration::SourceStrategy;
+use uv_distribution_types::{IndexLocations, Requirement};
+use uv_normalize::{DEV_DEPENDENCIES, GroupName, PackageName};
+use uv_workspace::dependency_groups::FlatDependencyGroups;
+use uv_workspace::pyproject::{Sources, ToolUvSources};
+use uv_workspace::{DiscoveryOptions, MemberDiscovery, VirtualProject, WorkspaceCache};
+
+use crate::metadata::{GitWorkspaceMember, LoweredRequirement, MetadataError};
+
+#[derive(Debug, Clone)]
+pub struct SourcedDependencyGroups {
+    pub name: Option<PackageName>,
+    pub dependency_groups: BTreeMap<GroupName, Box<[Requirement]>>,
+}
+
+impl SourcedDependencyGroups {
+    /// Lower by considering `tool.uv` in `pyproject.toml` if present, used for Git and directory
+    /// dependencies.
+    pub async fn from_virtual_project(
+        pyproject_path: &Path,
+        git_member: Option<&GitWorkspaceMember<'_>>,
+        locations: &IndexLocations,
+        source_strategy: SourceStrategy,
+        cache: &WorkspaceCache,
+    ) -> Result<Self, MetadataError> {
+        let discovery = DiscoveryOptions {
+            stop_discovery_at: git_member.map(|git_member| {
+                git_member
+                    .fetch_root
+                    .parent()
+                    .expect("git checkout has a parent")
+                    .to_path_buf()
+            }),
+            members: match source_strategy {
+                SourceStrategy::Enabled => MemberDiscovery::default(),
+                SourceStrategy::Disabled => MemberDiscovery::None,
+            },
+        };
+        let project = VirtualProject::discover_defaulted(pyproject_path, &discovery, cache).await?;
```

Comment:
Should this be included in the workspace discovery caching, could this reparse our entire workspace?

Reply:
ah it already has the right cache